bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/handler/LeapBridgeHandler.java

```diff
@@ -0,0 +1,763 @@
+            reader = new BufferedReader(new InputStreamReader(sslsocket.getInputStream()));
+            this.sslsocket = sslsocket;
+        } catch (UnknownHostException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Unknown host");
+            return;
+        } catch (IllegalArgumentException e) {
+            // port out of valid range
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Invalid port number");
+            return;
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Error opening SSL connection. Check log.");
+            logger.info("Error opening SSL connection: {}", e.getMessage());
+            disconnect();
+            scheduleConnectRetry(reconnectInterval); // Possibly a temporary problem. Try again later.
+            return;
+        }
+
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, STATUS_INITIALIZING);
+
+        Thread readerThread = new Thread(this::readerThreadJob, "Lutron reader");
+        readerThread.setDaemon(true);
+        readerThread.start();
+        this.readerThread = readerThread;
+
+        Thread senderThread = new Thread(this::senderThreadJob, "Lutron sender");
+        senderThread.setDaemon(true);
+        senderThread.start();
+        this.senderThread = senderThread;
+
+        sendCommand(new LeapCommand(Request.getButtonGroups()));
+        queryDiscoveryData();
+        sendCommand(new LeapCommand(Request.subscribeOccupancyGroupStatus()));
+
+        logger.debug("Starting keepAlive job with interval {}", heartbeatInterval);
+        keepAlive = scheduler.scheduleWithFixedDelay(this::sendKeepAlive, heartbeatInterval, heartbeatInterval,
+                TimeUnit.MINUTES);
+    }
+
+    /**
+     * Called by connect() and discovery service to request fresh discovery data
+     */
+    public void queryDiscoveryData() {
+        sendCommand(new LeapCommand(Request.getDevices()));
+        sendCommand(new LeapCommand(Request.getAreas()));
+        sendCommand(new LeapCommand(Request.getOccupancyGroups()));
+    }
+
+    private void scheduleConnectRetry(long waitMinutes) {
+        logger.debug("Scheduling connection retry in {} minutes", waitMinutes);
+        connectRetryJob = scheduler.schedule(this::connect, waitMinutes, TimeUnit.MINUTES);
+    }
+
+    private synchronized void disconnect() {
+        logger.debug("Disconnecting");
+
+        Thread senderThread = this.senderThread;
+        Thread readerThread = this.readerThread;
+
+        if (connectRetryJob != null) {
+            connectRetryJob.cancel(true);
+        }
+        if (keepAlive != null) {
+            keepAlive.cancel(true);
+        }
+
+        // May be called from keepAliveReconnect thread, so call cancel with false
+        reconnectTaskCancel(false);
```

Comment:
disconnect is called as part of `dispose` so it would probably be a really good idea to cancel all tasks with interrupts.

Reply:
Yes. I can't just call cancel with interrupt on the reconnect job all of the time, though. The reconnect job calls disconnect(), so it would shoot itself. I'll have to add a boolean parameter to disconnect() that gets set when called from dispose().

Reply:
Well interrupts are optional, you just need to make sure that the asynchronous task results are silently discarded.

Reply:
I've changed this to use reconnectTaskCancel(true) when called from dispose().